<a href="https://colab.research.google.com/github/Tonmf/ProyectoIML/blob/master/Casos/Caso_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Reporte de MLT de Barcelona Telco

## Caso 5: Fábrica Milwaukee (aka Bosch)


## Data

Source: https://www.kaggle.com/c/bosch-production-line-performance

In [ ]:
import os              # Package to use directory command to list files
import numpy as np     # linear algebra
import pandas as pd    # data processing, CSV file I/O (e.g. pd.read_csv)
import seaborn as sns  # Used for plotting the graph
import matplotlib.pyplot as plt
import re

from scipy import stats # for Statistics
from statistics import mean
from sklearn import preprocessing
from sklearn.metrics import accuracy_score
from scipy.stats import norm
from sklearn.preprocessing import StandardScaler
from matplotlib import rcParams
%matplotlib inline

In [ ]:
from google.colab import drive 
drive.mount('/content/gdrive')

In [5]:
data = pd.read_csv('/content/gdrive/My Drive/IML/Proyecto Final/train_numeric.csv',nrows=4000)

## Data Exploration

In [6]:
print(data.shape)
data.head(5)

(4000, 970)


Id  L0_S0_F0  L0_S0_F2  ...  L3_S51_F4260  L3_S51_F4262  Response
0   4     0.030    -0.034  ...           NaN           NaN         0
1   6       NaN       NaN  ...           NaN           NaN         0
2   7     0.088     0.086  ...           NaN           NaN         0
3   9    -0.036    -0.064  ...           NaN           NaN         0
4  11    -0.055    -0.086  ...           NaN           NaN         0

[5 rows x 970 columns]

In [ ]:
data.describe()

Id     L0_S0_F0  ...  L3_S51_F4262    Response
count  4000.000000  2325.000000  ...    194.000000  4000.00000
mean   4018.952500    -0.001055  ...      0.000196     0.00575
std    2307.764835     0.078488  ...      0.001686     0.07562
min       4.000000    -0.291000  ...      0.000000     0.00000
25%    2058.000000    -0.055000  ...      0.000000     0.00000
50%    4056.000000     0.003000  ...      0.000000     0.00000
75%    6030.250000     0.056000  ...      0.000000     0.00000
max    8000.000000     0.278000  ...      0.017000     1.00000

[8 rows x 970 columns]

In [ ]:
data.columns

Index(['Id', 'L0_S0_F0', 'L0_S0_F2', 'L0_S0_F4', 'L0_S0_F6', 'L0_S0_F8',
       'L0_S0_F10', 'L0_S0_F12', 'L0_S0_F14', 'L0_S0_F16',
       ...
       'L3_S50_F4245', 'L3_S50_F4247', 'L3_S50_F4249', 'L3_S50_F4251',
       'L3_S50_F4253', 'L3_S51_F4256', 'L3_S51_F4258', 'L3_S51_F4260',
       'L3_S51_F4262', 'Response'],
      dtype='object', length=970)

In [ ]:
data.dtypes

Id                int64
L0_S0_F0        float64
L0_S0_F2        float64
L0_S0_F4        float64
L0_S0_F6        float64
                 ...   
L3_S51_F4256    float64
L3_S51_F4258    float64
L3_S51_F4260    float64
L3_S51_F4262    float64
Response          int64
Length: 970, dtype: object

In [7]:
totalt = data.isnull().sum().sort_values(ascending=False)
percent = (data.isnull().sum()/data.isnull().count()).sort_values(ascending=False)
missing_data = pd.concat([totalt, percent], axis=1, keys=['Total', 'Percent'])
missing_data

Total  Percent
L1_S25_F2181   3997  0.99925
L1_S25_F2184   3997  0.99925
L1_S25_F2187   3997  0.99925
L1_S25_F2190   3997  0.99925
L1_S25_F2193   3997  0.99925
...             ...      ...
L3_S37_F3946    204  0.05100
L3_S37_F3948    204  0.05100
L3_S37_F3950    204  0.05100
Response          0  0.00000
Id                0  0.00000

[970 rows x 2 columns]

In [8]:
missing2 = missing_data.loc[(missing_data['Percent']>0.6)]
missing2

Total  Percent
L1_S25_F2181   3997  0.99925
L1_S25_F2184   3997  0.99925
L1_S25_F2187   3997  0.99925
L1_S25_F2190   3997  0.99925
L1_S25_F2193   3997  0.99925
...             ...      ...
L0_S7_F138     2832  0.70800
L0_S7_F142     2832  0.70800
L0_S7_F136     2832  0.70800
L0_S5_F114     2811  0.70275
L0_S5_F116     2811  0.70275

[796 rows x 2 columns]

In [ ]:
796/970 *100

82.06185567010309

In [9]:
missing2.index

Index(['L1_S25_F2181', 'L1_S25_F2184', 'L1_S25_F2187', 'L1_S25_F2190',
       'L1_S25_F2193', 'L1_S25_F2196', 'L1_S25_F2199', 'L1_S25_F2202',
       'L1_S25_F2682', 'L1_S25_F2707',
       ...
       'L0_S2_F44', 'L0_S2_F52', 'L0_S2_F56', 'L0_S2_F60', 'L0_S2_F64',
       'L0_S7_F138', 'L0_S7_F142', 'L0_S7_F136', 'L0_S5_F114', 'L0_S5_F116'],
      dtype='object', length=796)

In [ ]:
df_agrup = data.groupby(by=['Response'])
df_agrup.count()

Id  L0_S0_F0  L0_S0_F2  ...  L3_S51_F4258  L3_S51_F4260  L3_S51_F4262
Response                            ...                                          
0         3977      2316      2316  ...           193           193           193
1           23         9         9  ...             1             1             1

[2 rows x 969 columns]

In [ ]:
23/4000 *100

0.575

De este recuento podemos sacar un par de cosas interesantes: 

1.   Sólo hay 23 casos en que tenemos una pieza mala
2.   De un vistazo rápido vemos que en cierta líneas hay más 1's que en otras.

In [27]:
df3 = df2.drop(labels=missing2.index,axis=1)

In [28]:
df3.head()

Id  L0_S0_F0  L0_S0_F2  ...  L3_S37_F3948  L3_S37_F3950  Response
0   4     0.030    -0.034  ...           0.0           0.0         0
1   6       NaN       NaN  ...           0.0           0.0         0
2   7     0.088     0.086  ...           0.0           0.0         0
3   9    -0.036    -0.064  ...           0.0           0.0         0
4  11    -0.055    -0.086  ...           0.0           0.0         0

[5 rows x 174 columns]

In [29]:
df_agrup3 = df3.groupby(by=['Response'])
df_agrup3.count()

Id  L0_S0_F0  L0_S0_F2  ...  L3_S37_F3946  L3_S37_F3948  L3_S37_F3950
Response                            ...                                          
0         3977      2316      2316  ...          3775          3775          3775
1           23         9         9  ...            21            21            21

[2 rows x 173 columns]

In [30]:
totalt = df3.isnull().sum().sort_values(ascending=False)
percent = (df3.isnull().sum()/df3.isnull().count()).sort_values(ascending=False)
missing_data3 = pd.concat([totalt, percent], axis=1, keys=['Total', 'Percent'])
missing_data3

Total  Percent
L3_S36_F3918   2113  0.52825
L3_S36_F3938   2113  0.52825
L3_S36_F3934   2113  0.52825
L3_S36_F3930   2113  0.52825
L3_S36_F3926   2113  0.52825
...             ...      ...
L3_S37_F3944    204  0.05100
L3_S37_F3946    204  0.05100
L3_S37_F3948    204  0.05100
Response          0  0.00000
Id                0  0.00000

[174 rows x 2 columns]

In [45]:
df4 = df3.fillna(df3.mean())

In [46]:
totalt = df4.isnull().sum().sort_values(ascending=False)
percent = (df4.isnull().sum()/df4.isnull().count()).sort_values(ascending=False)
missing_data4 = pd.concat([totalt, percent], axis=1, keys=['Total', 'Percent'])
missing_data4

Total  Percent
Response          0      0.0
L3_S29_F3439      0      0.0
L3_S29_F3467      0      0.0
L3_S29_F3464      0      0.0
L3_S29_F3461      0      0.0
...             ...      ...
L3_S30_F3709      0      0.0
L3_S30_F3704      0      0.0
L3_S30_F3699      0      0.0
L3_S30_F3694      0      0.0
Id                0      0.0

[174 rows x 2 columns]

## Algoritmo de Machine Learning

In [50]:
missing_data4.index

Index(['Response', 'L3_S29_F3439', 'L3_S29_F3467', 'L3_S29_F3464',
       'L3_S29_F3461', 'L3_S29_F3458', 'L3_S29_F3455', 'L3_S29_F3452',
       'L3_S29_F3449', 'L3_S29_F3442',
       ...
       'L3_S30_F3734', 'L3_S30_F3729', 'L3_S30_F3724', 'L3_S30_F3719',
       'L3_S30_F3714', 'L3_S30_F3709', 'L3_S30_F3704', 'L3_S30_F3699',
       'L3_S30_F3694', 'Id'],
      dtype='object', length=174)

In [51]:
from sklearn.preprocessing import StandardScaler
target = df4['Response'].values

select_features = missing_data4.index


scaler = StandardScaler()
dfScaled = scaler.fit_transform(df4[select_features])

In [53]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(dfScaled, target, test_size=0.2, random_state=42)

In [54]:
from sklearn.ensemble import RandomForestClassifier  # importing model to use for our prediction
from sklearn.metrics import classification_report

RFC = RandomForestClassifier(max_depth = 5, min_samples_split = 4, n_estimators = 500,random_state = 20, n_jobs = -1)
RFC.fit(x_train, y_train) 
train_prediction = RFC.predict(x_train)
print("La accuracy es de {}.".format(accuracy_score(y_train,train_prediction)))
print(classification_report(y_train, train_prediction))

test_prediction = RFC.predict(x_test)
print("La accuracy es de {}.".format(accuracy_score(y_test, test_prediction)))
print(classification_report(y_test, test_prediction))

La accuracy es de 0.9959375.
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      3179
           1       1.00      0.38      0.55        21

    accuracy                           1.00      3200
   macro avg       1.00      0.69      0.77      3200
weighted avg       1.00      1.00      1.00      3200

La accuracy es de 0.99875.
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       798
           1       1.00      0.50      0.67         2

    accuracy                           1.00       800
   macro avg       1.00      0.75      0.83       800
weighted avg       1.00      1.00      1.00       800



## Resultados

El resultado es sorprendentemente alto. Cabe recordar que sólo teníamos un 0.05% de casos en que teníamos pieza mala, por lo que accuracy menor de 99,95 no es del todo interesante ya que en la mayoría de casos si predice pieza buena la acierta... Probamos en equilibrar el número de piezas buenas y piezas malas del dataset.

Como sólo teníamos 23 casos de pieza mala, vamos a cargar más datos para poder generar un dataset suficientemente grande.

In [134]:
df5 = pd.read_csv('/content/gdrive/My Drive/IML/Proyecto Final/train_numeric.csv',nrows=100000)

In [135]:
df_agrup = df5.groupby(by=['Response'])
df_agrup.count()

Id  L0_S0_F0  L0_S0_F2  ...  L3_S51_F4258  L3_S51_F4260  L3_S51_F4262
Response                             ...                                          
0         99432     57625     57625  ...          5171          5171          5171
1           568       290       290  ...            24            24            24

[2 rows x 969 columns]

In [136]:
568/100000*100

0.5680000000000001

In [137]:
df6 = df5.drop(labels=missing2.index,axis=1)

In [138]:
df7 = df6.fillna(df6.mean())

In [139]:
totalt = df7.isnull().sum().sort_values(ascending=False)
percent = (df7.isnull().sum()/df7.isnull().count()).sort_values(ascending=False)
missing_data4 = pd.concat([totalt, percent], axis=1, keys=['Total', 'Percent'])
missing_data4

Total  Percent
Response          0      0.0
L3_S29_F3439      0      0.0
L3_S29_F3467      0      0.0
L3_S29_F3464      0      0.0
L3_S29_F3461      0      0.0
...             ...      ...
L3_S30_F3709      0      0.0
L3_S30_F3704      0      0.0
L3_S30_F3699      0      0.0
L3_S30_F3694      0      0.0
Id                0      0.0

[174 rows x 2 columns]

In [140]:
df0 = df7.loc[(df7['Response'] == 1)]
df0.shape

(568, 174)

In [141]:
df1 = df7.loc[(df7['Response'] == 0)]
df1.shape

(99432, 174)

In [142]:
df1=df1[0:568]
df1.shape

(568, 174)

In [143]:
frames = [df1, df0]
df8 = pd.concat(frames)
df8.shape

(1136, 174)

### Algoritmo equilibrando Response

In [144]:
from sklearn.preprocessing import StandardScaler
target = df8['Response'].values

select_features = missing_data4.index


scaler = StandardScaler()
dfScaled = scaler.fit_transform(df8[select_features])

In [145]:
from sklearn.model_selection import train_test_split
x_train1, x_test1, y_train1, y_test1 = train_test_split(dfScaled, target, test_size=0.2, random_state=42)

In [146]:
from sklearn.ensemble import RandomForestClassifier  # importing model to use for our prediction
from sklearn.metrics import classification_report

RFC1 = RandomForestClassifier(max_depth = 5, min_samples_split = 4, n_estimators = 500,random_state = 20, n_jobs = -1)
RFC1.fit(x_train1, y_train1) 
train_prediction1 = RFC1.predict(x_train1)
print("La accuracy es de {}.".format(accuracy_score(y_train1,train_prediction1)))
print(classification_report(y_train1, train_prediction1))

test_prediction1 = RFC1.predict(x_test1)
print("La accuracy es de {}.".format(accuracy_score(y_test1, test_prediction1)))
print(classification_report(y_test1, test_prediction1))

La accuracy es de 1.0.
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       444
           1       1.00      1.00      1.00       464

    accuracy                           1.00       908
   macro avg       1.00      1.00      1.00       908
weighted avg       1.00      1.00      1.00       908

La accuracy es de 0.9956140350877193.
              precision    recall  f1-score   support

           0       0.99      1.00      1.00       124
           1       1.00      0.99      1.00       104

    accuracy                           1.00       228
   macro avg       1.00      1.00      1.00       228
weighted avg       1.00      1.00      1.00       228



En este caso teníamos un 0.56% de pieza mala. Por lo tanto buscamos una accuracy más alta que un 99.44%.

La accuracy en entrenamiento es del 100% por lo que podríamos concluir que hemos **sobreentrenado** el dataset. Pero en test tenemos una accuracy de **99.56%**, superior a lo que buscábamos.

 ## Discusión y Conclusiones

## Linias de desarrollo futuro

### Annexo 1


### Annexo 2